providers/implementations/keymgmt/ec_kmgmt.c

```diff
@@ -0,0 +1,253 @@
+            goto err;
+
+        ok = 1;
+    }
+
+ err:
+    OPENSSL_free(curve_name);
+    if (!ok) {
+        EC_GROUP_free(ecg);
+        ecg = NULL;
+    }
+    return ecg;
+}
+
+static ossl_inline
+int domparams_to_params(const EC_GROUP *ecg, OSSL_PARAM_BLD *tmpl)
+{
+    int curve_nid;
+
+    if (ecg == NULL)
+        return 0;
+
+    curve_nid = EC_GROUP_get_curve_name(ecg);
+
+    if (curve_nid == NID_undef) {
+        /* explicit parameters */
+
+        /*
+         * TODO(3.0): should we support explicit parameters curves?
+         */
+        return 0;
+    } else {
+        /* named curve */
+        const char *curve_name = NULL;
+
+        if ((curve_name = OBJ_nid2sn(curve_nid)) == NULL)
+            return 0;
+
+        if (!ossl_param_bld_push_utf8_string(tmpl, OSSL_PKEY_PARAM_EC_NAME, curve_name, 0))
+            return 0;
+    }
+
+    return 1;
+}
+
+static ossl_inline
+EC_KEY *params_to_key(const OSSL_PARAM params[])
+{
+    const OSSL_PARAM *param_priv_key, *param_pub_key;
+    BIGNUM *priv_key = NULL;
+    char *pub_key = NULL;
+    EC_GROUP *ecg = NULL;
+    EC_POINT *pub_point = NULL;
+    EC_KEY *ret = NULL;
+    int ok = 0;
+
+    if ((ecg = params_to_domparams(params)) == NULL)
+        return NULL;
+
+    param_priv_key =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_EC_PRIV_KEY);
+    param_pub_key =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_EC_PUB_KEY);
+
+    /*
+     * We want to have at least a public key either way, so we end up
+     * requiring it unconditionally.
+     */
+    if (param_pub_key == NULL)
+        goto err;
+
+    if ((param_priv_key != NULL
+                && !OSSL_PARAM_get_BN(param_priv_key, &priv_key))
+            || !OSSL_PARAM_get_utf8_string(param_pub_key, &pub_key, 0))
+        goto err;
+
+    if (!EC_POINT_hex2point(ecg, pub_key, pub_point, NULL))
+        goto err;
+
+    if ((ret = EC_KEY_new()) == NULL
+            || !EC_KEY_set_group(ret, ecg)
+            || !EC_KEY_set_private_key(ret, priv_key)
+            || !EC_KEY_set_public_key(ret, pub_point))
+        goto err;
+
+    ok = 1;
+
+ err:
+    BN_clear_free(priv_key);
+    OPENSSL_free(pub_key);
+    EC_POINT_free(pub_point);
+    EC_GROUP_free(ecg);
+    if (!ok) {
+        EC_KEY_free(ret);
+        ret = NULL;
+    }
+    return ret;
+}
+
+static ossl_inline
+int key_to_params(const EC_KEY *eckey, OSSL_PARAM_BLD *tmpl)
+{
+    const BIGNUM *priv_key = NULL;
+    const EC_POINT *pub_point = NULL;
+    const EC_GROUP *ecg = NULL;
+    char *pub_key = NULL;
+    int ret = 0;
+
+    if (eckey == NULL)
+        return 0;
+
+    ecg = EC_KEY_get0_group(eckey);
+    priv_key = EC_KEY_get0_private_key(eckey);
+    pub_point = EC_KEY_get0_public_key(eckey);
+
+    /* group and public_key must be present, priv_key is optional */
+    if (ecg == NULL || pub_point == NULL)
+        return 0;
+#define PNT_FORM POINT_CONVERSION_UNCOMPRESSED
```

Comment:
Why this?

Reply:
Oh, this was just leftover after experimenting around with different alternatives.

I will clean it up moving forward.